### ISSUE TITLE
Bottom blocks flickering while scrolling the page (iOS)

### ISSUE THREAD

**Issue by Prooksius** [Telegram-Mini-Apps/issues#50]
## Platform

- [ ] Telegram for Android
- [x] Telegram for iOS
- [ ] Telegram for macOS
- [ ] Telegram Desktop
- [ ] Telegram Web A
- [ ] Telegram Web K

## Describe the bug

This bug is only visible in iOS, iPhone.
It is completely gone after I swipe mini app down and reopen it.
Also it is gone after MainButton (at the bottom) showing. When it is shown - bug disappears untill I close and reopen the app.
It is gone after I change the orientation of the phone to the horizontal and turn it back to the vertical.

https://github.com/user-attachments/assets/59bac3fe-221d-4a7f-a565-f7245c885c48

**Comment by heyqbnk** [comment]
Is there any chance you are using position fixed or position absolute container and scrolling inside it?

**Comment by Prooksius** [comment]
No, static and relative. I don't use fixed and absolute app containers.
But when I tried to resolve this flickering, I tried to use fixed and absolute too. But that didn't help.
The only thing that helps - i show MainButton for half a second and close it while starting my app. Bug is gone!

**Comment by Prooksius** [comment]
Also there are no hidden blocks there that might overlap the visible. No any. The simple and clear markup.

**Comment by GoGeek91** [comment]
Hey guys, i am also facing that issue... can any one help me out.

**Comment by Prooksius** [comment]
Looks like we use some sort of styling that leads to such a behavior. Not everyone faces this issue.

As I wrote earlier, the only thing that helps me - I show MainButton for half a second and close it while starting my app. Bug is gone! 
Ugly, I agree, but it helps.

**Comment by heyqbnk** [comment]
Someone said, that this bug vanishes if you open the app using the direct link instead of Menu Button or the Open App button inside the bot info.

Could someone verify?

**Comment by Prooksius** [comment]
> Someone said, that this bug vanishes if you open the app using the direct link instead of Menu Button or the Open App button inside the bot info.
> 
> Could someone verify?

Yes, I can confirm it, tested a minute ago. App launched via link does not have this bug.

**Comment by heyqbnk** [comment]
Okay, waiting for some more upvotes and then passing our findings to the team.

Please, make sure that you can reproduce the problem when running the mini app not using the direct link, and you can't opening it via the direct one

**Comment by heyqbnk** [comment]
This issue was passed to the Telegram developers

**Comment by imaai** [comment]
@Prooksius  @heyqbnk I'm wondering how did you achieve that bottom menu not jumping up when reaching end of the page?

On iOS additional safe area padding is added, which causes overscroll of the whole document. 

Here you can find an bug demo: https://bugs.telegram.org/c/41529. 

I couldn't find any workaround that doesn't cause any other side effects. I mean except showing Secondary or Main button on every view 😆 

@heyqbnk Could we ping Telegram developers on that issue too? We have safe area insets in the WebApp API. It's much better to handle one padding from variable than fighting overflowing view that you can't control.

**Comment by imaai** [comment]
@heyqbnk By the way there is a sticky app guide in the docs: https://docs.telegram-mini-apps.com/platform/sticky-app 

This approach also has a bug. When you scroll over a long list with momentum and re-render page in the meantime (i.e navigate to different view with smaller size). You get scroll offset from the moon and basically you can see nothing + it's not scrollable.

This only happens when using native trigger i.e. BackButton. Regular touch events cause scroll to stop so there is no issue.

**Comment by Prooksius** [comment]
@imaai 
https://github.com/Telegram-Mini-Apps/issues/issues/39#issuecomment-2322866986 - try this method

**Comment by imaai** [comment]
@Prooksius Yeah I have already tried that.

Are you're doing some additional magic with reattaching scroll events?

This approach has consequences:  

- No longer expands, minimizes app when touching inside scrollable elements when reaching top. (This can be worked around)
- As mentioned above you are getting bugs with scrolling position when re-rendering and changing size of the document.

> [@heyqbnk](https://github.com/heyqbnk) By the way there is a sticky app guide in the docs: https://docs.telegram-mini-apps.com/platform/sticky-app
> 
> This approach also has a bug. When you scroll over a long list with momentum and re-render page in the meantime (i.e navigate to different view with smaller size). You get scroll offset from the moon and basically you can see nothing + it's not scrollable.
> 
> This only happens when using native trigger i.e. BackButton. Regular touch events cause scroll to stop so there is no issue.

I assume mine, yours issues all have the same underlying cause. I would check if you have the same flickering but with WebApp.MainButton or WebApp.SecondaryButton rendered. Cause this is as far as I know the only thing that locks properly document size and doesn't add additional safe area on iOS.

**Comment by Prooksius** [comment]
> Are you're doing some additional magic with reattaching scroll events?

no, nothing magical..

> I would check if you have the same flickering but with WebApp.MainButton or WebApp.SecondaryButton rendered.

No, with MainButton shown this bug disappears. Even if you hide it back.

> I assume mine, yours issues all have the same underlying cause.

I agree, looks like the same reason.

**Comment by artpopov** [comment]
I figured out that the bug appears if you open the mini app in compact mode and then call expand after opening. It feels like the app still "thinks" we are in compact mode and the flickering of the elements at the bottom seems to be part of the browser's optimization — a way to hide what hasn't made it into the compact viewport.

As a workaround, I guess we can try the following approaches:

1) Set position: fixed for html and body.
2) Trigger a reflow after expanding (for example, by accessing element.offsetHeight).
3) Enable GPU acceleration on scrollable elements (using will-change or transform: translateZ(0)).

**Comment by thepetruha** [comment]
`transform: translateZ(0) `
Is's working for me!

**Comment by jojudre** [comment]
Set position: fixed to body and width: 100% helped

**Comment by heyqbnk** [comment]
The problem was fixed in the latest update. Please, don't use position fixed on global containers (for body, for example) as it breaks the native web client behavior.
